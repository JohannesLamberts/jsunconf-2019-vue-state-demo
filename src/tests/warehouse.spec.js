import { createLocalVue } from '@vue/test-utils'
import StatePlugin from '@/plugins/state'
import WarehouseModel from '@/models/warehouse'
import DataApiModel from '@/models/dataApi'
import { serializeHistory } from '@/tests/vue-history-testing'

const Vue = createLocalVue()
Vue.use(StatePlugin)

describe('Warehouse', () => {
  let vm
  let Warehouse
  let DataApi

  beforeEach(async () => {
    vm = new Vue({
      models: {
        DataApi: DataApiModel,
        Warehouse: WarehouseModel,
      },
    })

    /* eslint-disable-next-line prefer-destructuring */
    DataApi = vm.DataApi
    /* eslint-disable-next-line prefer-destructuring */
    Warehouse = vm.Warehouse

    // wait for warehouse fetch to be done
    await Promise.all(vm.$globalHistory.items.map(item => item.promise))

    Warehouse.saveItems([
      {
        id: 'mocked',
        version: 2,
        count: 3,
      },
    ])

    // clear initial history
    vm.$globalHistory.items = []
  })

  it('should should synchronize items with external database', async () => {
    DataApi.setMode(200)
    await Warehouse.incoming(Warehouse.items[0].id)
    expect(serializeHistory(vm.$globalHistory))
      .toMatchSnapshot()
  })

  it('should synchronize and redo after remote change', async () => {
    DataApi.setMode(409)
    await Warehouse.incoming(Warehouse.items[0].id)
    expect(serializeHistory(vm.$globalHistory))
      .toMatchSnapshot()
  })

  it('should synchronize remotely removed items', async () => {
    DataApi.setMode(404)
    await Warehouse.incoming(Warehouse.items[0].id)
    expect(serializeHistory(vm.$globalHistory))
      .toMatchSnapshot()
  })
})
