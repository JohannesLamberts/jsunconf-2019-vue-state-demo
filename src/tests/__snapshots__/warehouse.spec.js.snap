// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Warehouse should should synchronize items with external database 1`] = `
"
DataApi.setMode(200)


Warehouse.incoming('mocked')
  
  Warehouse.modifyItemCount('mocked', 1)
      
    Warehouse.syncSafe({\\"id\\":\\"mocked\\",\\"version\\":2,\\"count\\":4}, null)
          
      DataApi.sync({\\"id\\":\\"mocked\\",\\"version\\":3,\\"count\\":4})
"
`;

exports[`Warehouse should synchronize and redo after remote change 1`] = `
"
DataApi.setMode(409)


Warehouse.incoming('mocked')
  
  Warehouse.modifyItemCount('mocked', 1)
      
    Warehouse.syncSafe({\\"id\\":\\"mocked\\",\\"version\\":2,\\"count\\":4}, null)
          
      DataApi.sync({\\"id\\":\\"mocked\\",\\"version\\":3,\\"count\\":4})
      ~~ errors with
      Remotely changed
              
        DataApi.setMode(200)
          
      Warehouse.updateItemFromRemote('mocked', {\\"status\\":409,\\"version\\":6,\\"count\\":900})
      
    Warehouse.modifyItemCount('mocked', 1)
          
      Warehouse.syncSafe({\\"id\\":\\"mocked\\",\\"version\\":6,\\"count\\":901}, null)
              
        DataApi.sync({\\"id\\":\\"mocked\\",\\"version\\":7,\\"count\\":901})
"
`;

exports[`Warehouse should synchronize remotely removed items 1`] = `
"
DataApi.setMode(404)


Warehouse.incoming('mocked')
  
  Warehouse.modifyItemCount('mocked', 1)
      
    Warehouse.syncSafe({\\"id\\":\\"mocked\\",\\"version\\":2,\\"count\\":4}, null)
          
      DataApi.sync({\\"id\\":\\"mocked\\",\\"version\\":3,\\"count\\":4})
      ~~ errors with
      Not found
          
      Warehouse.removeItem('mocked')
"
`;
